spinoso-time/src/time/tzrs/offset.rs

```diff
@@ -0,0 +1,273 @@
+            "Z" => Self::utc(),
+            _ => {
+                lazy_static! {
+                    static ref HH_MM_MATCHER: Regex = Regex::new(r"^([\-\+]{1})(\d{2})(\d{2})$").unwrap();
+                }
+                if HH_MM_MATCHER.is_match(input) {
+                    let caps = HH_MM_MATCHER.captures(input).unwrap();
+
+                    let sign = if caps.get(1).unwrap().as_str() == "+" { 1 } else { -1 };
+                    let hours = caps.get(2).unwrap().as_str().parse::<i32>().unwrap();
+                    let minutes = caps.get(3).unwrap().as_str().parse::<i32>().unwrap();
+
+                    let offset_seconds: i32 = sign * ((hours * SECONDS_IN_HOUR) + (minutes * SECONDS_IN_MINUTE));
+                    Self::fixed(offset_seconds)
+                } else {
+                    // TODO: ArgumentError
+                    Self::utc()
+                }
+            }
+        }
+    }
+}
+
+impl From<String> for Offset {
+    fn from(input: String) -> Self {
+        Offset::from(input.as_str())
+    }
+}
+
+impl From<TimeZoneRef<'static>> for Offset {
+    #[inline]
+    #[must_use]
+    fn from(tz: TimeZoneRef<'static>) -> Self {
+        Self::tz(tz)
+    }
+}
+
+impl From<i32> for Offset {
+    /// Construct a Offset with the offset in seconds from UTC
+    #[inline]
+    #[must_use]
+    fn from(seconds: i32) -> Self {
+        Self::fixed(seconds)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn offset_seconds_from_fixed_offset(input: &str) -> i32 {
+        let offset = Offset::from(input);
+        let local_time_type = offset.time_zone_ref().local_time_types()[0];
+        local_time_type.ut_offset()
+    }
+
+    fn offset_name(offset: &Offset) -> &str {
+        match offset {
+            Offset::Utc => "UTC",
+            Offset::Fixed(ltt) => ltt[0].time_zone_designation(),
+            Offset::Tz(_) => "Ambiguous timezone name",
+        }
+    }
+
+    #[test]
+    fn fixed_zero_is_not_utc() {
+        let offset = Offset::from(0);
+        assert!(matches!(offset, Offset::Fixed(_)));
+    }
+
+    #[test]
+    fn z_is_utc() {
+        let offset = Offset::from("Z");
+        assert!(matches!(offset, Offset::Utc));
+    }
+
+    #[test]
+    fn from_str_hh_mm() {
+        assert_eq!(0, offset_seconds_from_fixed_offset("+0000"));
+        assert_eq!(0, offset_seconds_from_fixed_offset("-0000"));
+        assert_eq!(60, offset_seconds_from_fixed_offset("+0001"));
+        assert_eq!(-60, offset_seconds_from_fixed_offset("-0001"));
+        assert_eq!(3600, offset_seconds_from_fixed_offset("+0100"));
+        assert_eq!(-3600, offset_seconds_from_fixed_offset("-0100"));
+        assert_eq!(7320, offset_seconds_from_fixed_offset("+0202"));
+        assert_eq!(-7320, offset_seconds_from_fixed_offset("-0202"));
+        assert_eq!(362_340, offset_seconds_from_fixed_offset("+9999"));
+        assert_eq!(-362_340, offset_seconds_from_fixed_offset("-9999"));
```

Comment:
It looks like MRI treats cases where `HH` or `MM` are out of range as errors.

```
[3.1.2] > Time.new(2022, 3, 12, 15, 23, 59, "+9900")
<internal:timev>:310:in `initialize': utc_offset out of range (ArgumentError)
        from (irb):14:in `new'
        from (irb):14:in `<main>'
        from /usr/local/var/rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `<main>'
[3.1.2] > Time.new(2022, 3, 12, 15, 23, 59, "+0199")
<internal:timev>:310:in `initialize': "+HH:MM", "-HH:MM", "UTC" or "A".."I","K".."Z" expected for utc_offset: +0199 (ArgumentError)
        from (irb):15:in `new'
        from (irb):15:in `<main>'
        from /usr/local/var/rbenv/versions/3.1.2/lib/ruby/gems/3.1.0/gems/irb-1.4.1/exe/irb:11:in `<top (required)>'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `load'
        from /usr/local/var/rbenv/versions/3.1.2/bin/irb:25:in `<main>'
```

Reply:
Thrown into the Errors PR since we wrap the errors in that one.